dbms/src/IO/Compression/CompressionCodecFOR.cpp

```diff
@@ -0,0 +1,254 @@
+UInt32 CompressionCodecFOR::compressData(const T * source, UInt32 count, char * dest)
+{
+    std::vector<T> values(count);
+    for (UInt32 i = 0; i < count; ++i)
+        values[i] = source[i];
+    T frame_of_reference = *std::min_element(values.cbegin(), values.cend());
+    // store frame of reference
+    unalignedStore<T>(dest, frame_of_reference);
+    dest += sizeof(T);
+    if (frame_of_reference != 0)
+    {
+        for (auto & value : values)
+            value -= frame_of_reference;
+    }
+    T max_value = *std::max_element(values.cbegin(), values.cend());
+    UInt8 width = BitpackingPrimitives::minimumBitWidth(max_value);
+    // store width
+    unalignedStore<UInt8>(dest, width);
+    dest += sizeof(UInt8);
+    // if width == 0, skip bitpacking
+    if (width == 0)
+        return sizeof(T) + sizeof(UInt8);
+    auto required_size = BitpackingPrimitives::getRequiredSize(count, width);
+    // after applying frame of reference, all values are bigger than 0.
+    BitpackingPrimitives::packBuffer(reinterpret_cast<unsigned char *>(dest), values.data(), count, width);
+    return sizeof(T) + sizeof(UInt8) + required_size;
+}
+
+template <std::integral T>
+void CompressionCodecFOR::decompressData(const char * source, UInt32 source_size, char * dest, UInt32 output_size)
+{
+    const auto count = output_size / sizeof(T);
+    T frame_of_reference = unalignedLoad<T>(source);
+    source += sizeof(T);
+    auto width = unalignedLoad<UInt8>(source);
+    source += sizeof(UInt8);
+    const auto required_size = source_size - sizeof(T) - sizeof(UInt8);
+    RUNTIME_CHECK(BitpackingPrimitives::getRequiredSize(count, width) == required_size);
+    auto round_size = BitpackingPrimitives::roundUpToAlgorithmGroupSize(count);
+    if (round_size != count)
+    {
+        // Reserve enough space for the temporary buffer.
+        unsigned char tmp_buffer[round_size * sizeof(T)];
+        BitpackingPrimitives::unPackBuffer<T>(
+            tmp_buffer,
+            reinterpret_cast<const unsigned char *>(source),
+            count,
+            width);
+        CompressionCodecFOR::applyFrameOfReference(reinterpret_cast<T *>(tmp_buffer), frame_of_reference, count);
+        memcpy(dest, tmp_buffer, output_size);
+        return;
+    }
+    BitpackingPrimitives::unPackBuffer<T>(
+        reinterpret_cast<unsigned char *>(dest),
+        reinterpret_cast<const unsigned char *>(source),
+        count,
+        width);
+    CompressionCodecFOR::applyFrameOfReference(reinterpret_cast<T *>(dest), frame_of_reference, count);
+}
+
+template <std::integral T>
+void CompressionCodecFOR::applyFrameOfReference(T * dst, T frame_of_reference, UInt32 count)
+{
+    if (!frame_of_reference)
+        return;
+
+    UInt32 i = 0;
+    UInt32 misaligned_count = count;
+#if defined(__AVX2__)
+    misaligned_count = count % (sizeof(__m256i) / sizeof(T));
+#endif
+    for (; i < misaligned_count; ++i)
+    {
+        dst[i] += frame_of_reference;
+    }
```

Comment:
I think doing it after the `for (; i < count; i += (sizeof(__m256i) / sizeof(T)))` can bring better performance. Because  we can assume the memory address of `dst + 0` should be aligned with the cache line, but it is not always aligned when it starts with `dst + misaligned_count`

Reply:
Good catch, fixed.